### ISSUE TITLE
Have presets for “Word characters”

### ISSUE THREAD

**Issue by GrimPixel** [jzohrab/lute-v3#220]
**Is your feature request related to a problem? Please describe.**

The “Word characters” requires some efforts to complete every time adding a language.

**Describe the solution you'd like**

Have presets for it, like “Latin”.

**Comment by jzohrab** [comment]
Hm, those are preset in the predefined languages.  I don't follow the request, can you give me an example?  Thanks @GrimPixel :wave:

**Comment by GrimPixel** [comment]
Hi👋
I see that it is often neglected by people to split the language and the writing system apart.. They are essentially independent from each other In my projects, To indicate a locale, I usually use ISO 639-3 alpha-3 and ISO 15924, and if needed, also ISO 3166-3 alpha-2. So it is like `eng-Latn` or `eng-Latn-US`.

I think it could be split here as well. Otherwise, each time I add a language, I have to check the Unicode charts to see if the characters are covered in the predefined range in another language.

**Comment by jzohrab** [comment]
ok, so each preset lang code would map to a character/regex range somehow?

**Comment by GrimPixel** [comment]
I think it could be like this: set up the language and the writing system separately: have “Writing Systems” alongside “Languages”. “Right-to-left” to “Word characters” are moved from “Languages” to “Writing  Systems”. The user creates a writing system first and a language second.

**Comment by M-Biggles** [comment]
> I think it could be like this: set up the language and the writing system separately: have “Writing Systems” alongside “Languages”. “Right-to-left” to “Word characters” are moved from “Languages” to “Writing Systems”. The user creates a writing system first and a language second.

That sounds rather convoluted and linguistically messy. Scripts are an afterthought and not the primary consideration. Some have several scripts and that's something we're working to implement. We started a project recently for adding new language presets.

You can check out our progress here:

https://docs.google.com/spreadsheets/d/1iBSwsKI-57HpjnNP8holmk_HDmjh16i5V_G8HJ-oNHY/edit#gid=0

It's a work in progress, and we will be improving each as we test more and get feedback from users.

**Comment by GrimPixel** [comment]
@M-Biggles I don't believe it is necessary to define dictionaries like this. There are so many dictionaries and I believe it needs to be left to the user to decide which one to choose: https://polyglotclub.com/wiki/Language/Multiple-languages/Culture/Internet-Dictionaries

If the writing system is dependent on language here, there are many questions to be answered: For a language, should its character range defined according to the alphabet? In English it's A-Za-z, but this will exclude any word that is capped and is English, like café. If é is included, what about á? What about ó? There is no end to this. So I think simply defining a writing system, to include all Latin letters, can solve this.

**Comment by M-Biggles** [comment]
> @M-Biggles I don't believe it is necessary to define dictionaries like this. There are so many dictionaries and I believe it needs to be left to the user to decide which one to choose: https://polyglotclub.com/wiki/Language/Multiple-languages/Culture/Internet-Dictionaries
> 
> If the writing system is dependent on language here, there are many questions to be answered: For a language, should its character range defined according to the alphabet? In English it's A-Za-z, but this will exclude any word that is capped and is English, like café. If é is included, what about á? What about ó? There is no end to this. So I think simply defining a writing system, to include all Latin letters, can solve this.

The user does have the ability to choose their own dictionary. It could be improved, but it's already possible.

English uses 'a-zA-ZÀ-ÖØ-öø-ȳáéíóúÁÉÍÓÚñÑ' for its Word Characters, not A-Za-z.

I feel like you've not really used the software yet. These things are possible and already implemented.

**Comment by jzohrab** [comment]
@GrimPixel — can you give me an example of what you’re describing for this feature?  Is it something like “latin-123” = “<some character set>”, and then I specify “latin-123” for the language?

**Comment by GrimPixel** [comment]
OK, there can be default dictionaries. I just use my own-picked ones.

If English is using that many characters, and it is for other Latin languages as well. It makes more sense to have the writing system separated instead of configuring every time for each language in Latin script.

**Comment by M-Biggles** [comment]
> OK, there can be default dictionaries. I just use my own-picked ones.
> 
> If English is using that many characters, and it is for other Latin languages as well. It makes more sense to have the writing system separated instead of configuring every time for each language in Latin script.

We will still need to define each language separately, but maybe we could create presets like:
[Latin] = 'a-zA-ZÀ-ÖØ-öø-ȳáéíóúÁÉÍÓÚñÑ'

**Comment by GrimPixel** [comment]
Let me explain my thoughts:
At first, I saw the “Word characters” in a text box instead of a dropdown menu, so I asked if there can be presets for “Word characters”. Then I have something new in my mind: why it is `a-zA-ZÀ-ÖØ-öø-ȳáéíóúÁÉÍÓÚñÑ` and not covering all Latin characters, as well as “Right-to-left” and other options all dedicated to a writing system, so I think having a separate “Writing Systems” in “Settings” can do.

**Comment by jzohrab** [comment]
Right, so it’s a reusable mapping to use for each language.  I’ll leave it in the backlog, it may tie in to some other things down the road. Thanks for the clarification.

**Comment by GrimPixel** [comment]
Glad to know this. Thank you both for listening.

**Comment by M-Biggles** [comment]
> Right, so it’s a reusable mapping to use for each language. I’ll leave it in the backlog, it may tie in to some other things down the road. Thanks for the clarification.

More of an ease-of-use thing.

I'll have a look into it and create some mappings and see which languages would be covered by each.

**Comment by GrimPixel** [comment]
I think it's very useful for specific languages. If a word is added in one writing system,  it will be available in all other writing systems. For example:
* Japanese: Japanese, Japanese syllabaries
* Kazakh: Cyrillic, Latin
* Mandarin Chinese: Han Simplified variant, Han Traditional variant, Bopomofo, Latin
* Mongolian: Cyrillic, Mongolian
